pennylane/capture/transforms/capture_cancel_inverses.py

```diff
@@ -0,0 +1,187 @@
+
+        Args:
+            op (Operator): a pennylane operator instance
+
+        Returns:
+            Any
+
+        This method is only called when the operator's output is a dropped variable,
+        so the output will not affect later equations in the circuit.
+
+        See also: :meth:`~.interpret_operation_eqn`.
+
+        """
+        # pylint: disable=too-many-branches
+        if len(op.wires) == 0:
+            return super().interpret_operation(op)
+
+        prev_op = self.previous_ops.get(op.wires[0], None)
+        if prev_op is None:
+            for w in op.wires:
+                self.previous_ops[w] = op
+            return []
+
+        cancel = False
+        if _are_inverses(op, prev_op):
+            # Same wires, cancel
+            if op.wires == prev_op.wires:
+                cancel = True
+            # Full overlap over wires
+            elif len(Wires.shared_wires([op.wires, prev_op.wires])) == len(op.wires):
+                # symmetric op + full wire overlap; cancel
+                if op in symmetric_over_all_wires:
+                    cancel = True
+                # symmetric over control wires, full overlap over control wires; cancel
+                elif op in symmetric_over_control_wires and (
+                    len(Wires.shared_wires([op.wires[:-1], prev_op.wires[:-1]]))
+                    == len(op.wires) - 1
+                ):
+                    cancel = True
+            # No or partial overlap over wires; can't cancel
+
+        if cancel:
+            for w in op.wires:
+                self.previous_ops.pop(w)
+            return []
+
+        # Putting the operations in a set to avoid applying the same op multiple times
+        # Using a set causes order to no longer be guaranteed, so the new order of the
+        # operations might differ from the original order. However, this only impacts
+        # operators without any shared wires, so correctness will not be impacted.
+        previous_ops_on_wires = set(self.previous_ops.get(w) for w in op.wires)
+        for o in previous_ops_on_wires:
+            if o is not None:
+                for w in o.wires:
+                    self.previous_ops.pop(w)
+        for w in op.wires:
+            self.previous_ops[w] = op
+
+        res = []
+        for o in previous_ops_on_wires:
+            res.append(super().interpret_operation(o))
+        return res
+
+    def interpret_all_previous_ops(self) -> None:
+        """Interpret all ops in ``previous_ops``. This is done whenever any
+        operators that haven't been interpreted that are saved to be cancelled
+        no longer need to be saved."""
+        ops_remaining = set(self.previous_ops.values())
+        for op in ops_remaining:
+            super().interpret_operation(op)
+
+        all_wires = tuple(self.previous_ops.keys())
+        for w in all_wires:
+            self.previous_ops.pop(w)
+
+    def eval(self, jaxpr: "jax.core.Jaxpr", consts: list, *args) -> list:
```

Comment:
I don't love how this is almost-but-not-quite the same thing as in the base class. Is handling of `previous_ops` something that we will need again for similar transforms? I think when we discussed a while ago, we concluded that it isn't that unusual that a transform uses information about previous operators to decide what to do.

I'm wondering if a version of `interpret_all_previous_ops` that per default does nothing should be part of the base class, and the base class `eval` should be updated to do this, instead of having an almost identical copy here with a few differences. But that seems like a task for refactoring after more transforms have been implemented and there's a more concrete understanding of how we might unify things, not something that should happen now.

Reply:
I had the exact same thought. I also think it might be worth breaking down the logic in `PlxprInterpreter.eval` into multiple smaller functions that will make it easier to override small chunks rather than having to override `eval`. I tried to make a `TransformInterpreter` base class locally with a "general" `interpret_previous_ops`, but I was concerned about introducing something too specific that might be worthwhile.

Since it only impacts one interpreter right now, I'm in favour of leaving it as is for now and refactoring the base class later as needed. As we add more interpreters, I'm sure we will come across more opportunities for improvements.